### PR TITLE
fix(settings): improve GQL error handling to avoid redirect loop

### LINKED ISSUE #6400: Don’t redirect endlessly when the GQL server can’t be reached
We direct to `get_flow` and back to settings in an endless loop when the "general application error" is shown. The page should instead stay in place at the general application error.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/FXA-2524)
┆Issue Number: FXA-2524

### PR DESCRIPTION
**Pull request by 6a68**
I'm annoyed by the branch name, but whatever. Final changes, from the commit message:

fix(settings): improve GQL error handling to avoid redirect loop

* Update the GQL API server to only return an AuthenticationError if it
    received an auth-related 400 or 401 error from auth-server. Otherwise,
    just return a generic ApolloError.

* Add tests for the front-end GQL library to be sure this works
    correctly.

* Add tests for the front-end GQL lib.

Fixes #6400.

-----

Original description from WIP draft:

The bug is that the gql-api was always returning an auth error, even if the problem was that it couldn't hit auth server. Setting up the (missing) tests for this lib has been a little complex; just looking for some feedback on test harness setup. Mocking out globals vs. wrapping in accessor functions might be a tradeoff that's been worked out elsewhere in new-settings?

### CODE REVIEW

File: packages/fxa-graphql-api/src/lib/auth.ts [hunk 1]
```diff
@@ -17,7 +17,16 @@ export class SessionTokenAuth {
     try {
       return await this.authClient.sessionStatus(sessionToken);
     } catch (err) {
-      throw new AuthenticationError('Invalid session token');
+      if (err.code === 400 || err.code === 401) {
```
Comment (6a68) [inline packages/fxa-graphql-api/src/lib/auth.ts:20]: auth client error codes are defined here: https://github.com/mozilla/fxa/blob/main/packages/fxa-auth-server/docs/api.md#defined-errors

Comment (bbangert) [reply packages/fxa-graphql-api/src/lib/auth.ts:20]: 400 is a bad request, maybe we should always throw ApolloError if we can customize it better?

Comment (6a68) [reply packages/fxa-graphql-api/src/lib/auth.ts:20]: The slight abuse of the 400 error code is pretty clear from the auth server docs that I linked above. Seems weird that "unverified account" throws a 400, not a 401, but here we are.

We could instead always throw ApolloError, but then we'll lose the "redirect on auth error" logic that we have on the front-end, or have to hyper-specialize it to various auth server `errno` values :-\

**Comment by 6a68** [comment]
I'm annoyed by the branch name, but whatever. Final changes, from the commit message:

-----

fix(settings): improve GQL error handling to avoid redirect loop

* Update the GQL API server to only return an AuthenticationError if it
received an auth-related 400 or 401 error from auth-server. Otherwise,
just return a generic ApolloError.

* The settings code handles auth errors by redirecting to /get_flow,
which then redirects to /signin if there's no session. But if the user
has already been redirected to /get_flow once, then there will be flow
params on the URL; if these params are present, don't redirect, just
show an error screen.

* Add tests for the front-end GQL lib.

Fixes #6400.

File: packages/fxa-settings/src/lib/gql.test.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    beforeEach(() => {
+      errorResponse = {
+        graphQLErrors: [
+          new GraphQLError(
+            'Incorrect password',
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
```
Comment (6a68) [inline packages/fxa-settings/src/lib/gql.test.ts:49]: I'm not a fan of this explicit passing of `undefined` thing. suggestions welcome on a nicer approach :-)

File: packages/fxa-settings/src/lib/gql.test.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    });
+  });
+
+  describe('if called with a 401 NetworkError', () => {
+    let networkError: any;
+
+    beforeEach(() => {
+      // Clumsily imitating the ServerError type :-\
+      networkError = new Error('Network error');
+      networkError.statusCode = 401;
```
Comment (6a68) [inline packages/fxa-settings/src/lib/gql.test.ts:90]: guess I should cast this to the ServerError type?

File: packages/fxa-settings/src/lib/gql.ts [hunk 4]
```diff
@@ -1,8 +1,32 @@
 import { ApolloClient, createHttpLink, from } from '@apollo/client';
 import { setContext } from '@apollo/client/link/context';
-import { onError } from '@apollo/client/link/error';
+import { ErrorHandler, onError } from '@apollo/client/link/error';
 import { cache, sessionToken, typeDefs } from './cache';
 
+export const errorHandler: ErrorHandler = ({ graphQLErrors, networkError }) => {
```
Comment (6a68) [inline packages/fxa-settings/src/lib/gql.ts:6]: extracted this out so I could unit test it

File: packages/fxa-settings/src/lib/gql.ts [hunk 5]
```diff
@@ -1,8 +1,32 @@
...
+        reauth = true;
+      }
+    }
+  }
+  if (networkError && 'statusCode' in networkError) {
+    if (networkError.statusCode === 401) {
+      reauth = true;
+    }
+  }
+  if (reauth && !hasFlowParams) {
```
Comment (6a68) [inline packages/fxa-settings/src/lib/gql.ts:21]: this `hasFlowParams` bit is the part that changed

File: packages/fxa-settings/package.json [hunk 6]
```diff
@@ -71,6 +71,7 @@
     "eslint-plugin-jest": "^23.13.2",
     "eslint-plugin-react": "7.19.0",
     "fxa-shared": "workspace:*",
+    "jest-mock-extended": "^1.0.10",
```
Comment (6a68) [inline packages/fxa-settings/package.json:74]: I'm definitely not sure this is actually needed, it just made mocking out the `ErrorResponse` arguments easier. Happy to use something else to keep our dependencies down 🤷

Comment (dannycoates) [reply packages/fxa-settings/package.json:74]: in this case since the mocks aren't being used/checked we should be able to just do this in the test file to satisfy the compiler:
```ts
const operation = null as any as Operation;
const forward = null as any as NextLink;
```

File: packages/fxa-graphql-api/src/lib/auth.ts (hunk 1 above)
Comment (dannycoates) [reply packages/fxa-graphql-api/src/lib/auth.ts:20]: This seems fine. The answer to any 4xx error on `/session/status` is reauth anyway

File: packages/fxa-settings/src/lib/gql.ts (hunk 5 above)
Comment (dannycoates) [reply packages/fxa-settings/src/lib/gql.ts:21]: I'm not sure `hasFlowParams` is the right thing to check, or if any addition here is necessary... iiuc the infinite loop was caused by the gql-api server not being able to access the auth-server, and it only sending auth errors. By returning an ApolloError for non-auth errors I think we've fixed that problem. I can imagine a situation where we get a legit auth error and do have flowParams where we still want to reauth instead of showing an error... like a page refresh after you deleted that session elsewhere

tldr: i think the gql server change is all we need/want

Comment (6a68) [reply packages/fxa-settings/src/lib/gql.ts:21]: Good point. We definitely might need to reauth when there are flow params. And I'm not sure the flow params should be in the query string anyway, which means this would need to be tweaked whenever that other change happens. I'll remove this bit from the PR.

Comment (6a68) [reply packages/fxa-settings/src/lib/gql.ts:21]: I'm still getting some weird reload loop if I nav to localhost:3000/beta/settings locally. Is that just me?

Comment (6a68) [reply packages/fxa-settings/src/lib/gql.ts:21]: lol never mind, I should be going to localhost:**3030**/beta/settings 🤦

**Review by dannycoates — APPROVED** [review]
a few red squiggles in gql.test.ts but r+ 👍
